File: JS_Submissions/linked_list/fast_and_slow_pointers/example3.js
/* 
Example 3: Given the head of a linked list and an integer k, return the kth node from the end.

For example, given the linked list that represents 1 -> 2 -> 3 -> 4 -> 5 and k = 2, return the node with value 4,
as it is the 2nd node from the end.

This problem is very similar to the first example. Again, we could just convert the list to an array, or we 
could iterate through once to find the length and then iterate again once we know the length, but there is a 
more elegant solution.

If we separate the two pointers by a gap of k, and then move them at the same speed, they will always be k apart.
When the fast pointer (the one further ahead) reaches the end, then the slow pointer must be at the desired node,
since it is k nodes behind.
*/

let findNode = (head, k) => {
    let slow = head;
    let fast = head;
    for (let i = 0; i < k; i++) {
        fast = fast.next;
    }

    while (fast) {
        slow = slow.next;
        fast = fast.next;
    }

    return slow;
}

// For the same reasons as in the first example, the time complexity of this algorithm is O(n) and the space
// complexity is O(1), where nn is the number of nodes in the linked list.

// Try solving these upcoming practice problems using the fast and slow pointer technique.

File: Rust_Submissions/dynamic_programming/example3.rs
/*
Example 3: 2140. Solving Questions With Brainpower

You are given a 0-indexed 2D integer array questions where questions[i] = [points_i, brainpower_i]
You have to process the questions in order. Solving question i will earn you points_i points but you will be 
unable to solve each of the next brainpower_i questions. If you skip question i, you get to decide on the next 
question. Return the maximum points you can score.


How can we tell this problem should be solved with DP? First, it is asking for a maximum score. Second, at every
question we need to make a decision: take or skip, and these decisions affect future decisions. If we decide to
take a question, it prevents us from taking some future questions.

As you may expect by now, we can define a function dp that returns the maximum score we can achieve. What 
information do we need at each state (other than an index variable i to indicate the current question we are on)?
We could include an integer that represents how many more questions we need to skip until we can start solving
questions again, but similar to with house robber, we can encode this information in our recurrence relation, 
so we'll just stick with dp(i) returning the maximum score.

If we are at the i^th question, we have two options:

Solve the question. We gain questions[i][0] points, but we cannot solve the next i questions. The next question 
we can solve is at index j = i + questions[i][1] + 1. Therefore, the total score is questions[i][0] + dp(j).
Skip the question. Like the problem says, this means we move on to the next question and make another decision 
there. The score is dp(i + 1).
This gives us our recurrence relation:

dp(i) = max(questions[i][0] + dp(j), dp(i + 1)), where j = i + questions[i][1] + 1

Since we are actually moving forward through the array instead of backward, this means the base case must be at 
the end. If i >= questions.length, that means the exam is over, so we can't score any more points. Therefore, 
the base case is:

dp(i) = 0, when i >= n
*/

fn most_points(questions: &Vec<(i32, usize)>) -> i32 {
    fn dp(i: usize, questions: &Vec<(i32, usize)>, memo: &mut Vec<i32>) -> i32 {
        if i >= questions.len() {
            return 0;
        }

        if memo[i] != -1 {
            return memo[i];
        }

        let j = i + questions[i].1 + 1;
        memo[i] = (questions[i].0 + dp(j, questions, memo)).max(dp(i + 1, questions, memo));
        memo[i]
    }

    let mut memo = vec![-1; questions.len()];
    dp(0, questions, &mut memo)
}

fn main() {
    let questions = vec![(3, 0), (5, 1), (6, 2), (4, 0), (2, 2)];
    println!("{}", most_points(&questions));
}

/*
There are O(n) states, where nn is the length of the input array, and each state costs O(1) to compute, 
once again giving us a time and space complexity of O(n)O(n). Because the recurrence relation is not static
(it depends on questions[i][1]), we cannot improve the space complexity in the bottom-up implementation.

*/